aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java

```diff
@@ -360,4 +416,28 @@ protected void finalize() throws Throwable {
       LOG.warn("Unclosed output stream created by:\n\t{}", trace);
     }
   }
+
+  private static class FileAndDigest {
+    private final File file;
+    private byte[] digest;
+    private boolean isDigestSet = false;
+
+    FileAndDigest(File file) {
+      this.file = file;
+    }
+
+    File file() {
+      return file;
+    }
+
+    void setDigest(byte[] digest) {
```

Comment:
Does this need to be mutable? Why not just pass in `digest` when creating? I thought that was possible before.

Reply:
It needs to be mutable if we are not passing MessageDigest. Earlier I had it mutable as I was passing MessageDigest, which I was able to update while bytes are written. File is created in newStream, but digest bytes are available later when stream is closed.

Reply:
In that case, I think it is a reason to go back to storing the digest. I didn't realize that the digest was getting updated after creating this. Seems like there should be a cleaner way where we update the digest and create an immutable object when the file is closed.

Reply:
+1 to Ryan's comment.  Keep a digest per file.  It's a trivial amount of memory and better encapsulated.